<a href="https://colab.research.google.com/github/Dat2784/Homelander/blob/main/docs/notebooks/Ames_Housing_Ver2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

df = pd.read_csv('/content/AmesHousing.csv')
display(df.head())

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot Shape Land Contour  ... Pool Area Pool QC  Fence Misc Feature  \
0   NaN       IR1          Lvl  ...         0     NaN    NaN          NaN   
1   NaN       Reg          Lvl  ...         0     NaN  MnPrv          NaN   
2   NaN       IR1          Lvl  ...         0     NaN    NaN         Gar2   
3   NaN       Reg          Lvl  ...         0     NaN    NaN          NaN   
4   NaN       IR1          Lvl  ...         0     NaN  MnPrv          NaN   

  Misc Val Mo Sold Yr Sold Sale Type  Sale Condition  SalePrice  
0        0       5    2010       WD           Normal     215000  
1        0       6    2010       WD           Normal     105000  
2    12500       6    2010       WD           Normal     172000  
3        0       4    2010       WD           Normal     244000  
4        0       3    2010       WD           Normal     189900  

[5 rows x 82 columns]

In [2]:
# Select categorical columns
categorical_cols = df.select_dtypes(include=['object', 'category']).columns

# Impute missing values in categorical columns with the mode
for col in categorical_cols:
    if df[col].isnull().any():
        mode_value = df[col].mode()[0]
        df[col].fillna(mode_value, inplace=True)

# Verify that missing values have been imputed
missing_values_after_categorical_imputation = df[categorical_cols].isnull().sum()
display(missing_values_after_categorical_imputation[missing_values_after_categorical_imputation > 0])

/tmp/ipython-input-2927222647.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(mode_value, inplace=True)


Series([], dtype: int64)

In [3]:
# Select numerical columns
numerical_cols = df.select_dtypes(include=['number']).columns

# Impute missing values in numerical columns with the mean
for col in numerical_cols:
    if df[col].isnull().any():
        mean_value = df[col].mean()
        df[col].fillna(mean_value, inplace=True)

# Verify that missing values have been imputed
missing_values_after_imputation = df[numerical_cols].isnull().sum()
display(missing_values_after_imputation[missing_values_after_imputation > 0])

/tmp/ipython-input-3753440199.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(mean_value, inplace=True)


Series([], dtype: int64)

In [5]:
missing_values = df.isnull().sum()
# Display columns with missing values
display(missing_values[missing_values > 0])

Series([], dtype: int64)

In [6]:
y = df['SalePrice']
X = df.drop('SalePrice', axis=1)

display(y.head())
display(X.head())

0    215000
1    105000
2    172000
3    244000
4    189900
Name: SalePrice, dtype: int64

Order        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street  \
0      1  526301100           20        RL         141.0     31770   Pave   
1      2  526350040           20        RH          80.0     11622   Pave   
2      3  526351010           20        RL          81.0     14267   Pave   
3      4  526353030           20        RL          93.0     11160   Pave   
4      5  527105010           60        RL          74.0     13830   Pave   

  Alley Lot Shape Land Contour  ... Screen Porch Pool Area Pool QC  Fence  \
0  Grvl       IR1          Lvl  ...            0         0      Ex  MnPrv   
1  Grvl       Reg          Lvl  ...          120         0      Ex  MnPrv   
2  Grvl       IR1          Lvl  ...            0         0      Ex  MnPrv   
3  Grvl       Reg          Lvl  ...            0         0      Ex  MnPrv   
4  Grvl       IR1          Lvl  ...            0         0      Ex  MnPrv   

  Misc Feature Misc Val Mo Sold Yr Sold  Sale Type  Sale Condition  
0         Shed        0       5    2010        WD           Normal  
1         Shed        0       6    2010        WD           Normal  
2         Gar2    12500       6    2010        WD           Normal  
3         Shed        0       4    2010        WD           Normal  
4         Shed        0       3    2010        WD           Normal  

[5 rows x 81 columns]

In [7]:
categorical_cols = X.select_dtypes(include=['object', 'category']).columns.tolist()
numerical_cols = X.select_dtypes(include=['number']).columns.tolist()

print("Categorical Columns:")
print(categorical_cols)
print("\nNumerical Columns:")
print(numerical_cols)

Categorical Columns:
['MS Zoning', 'Street', 'Alley', 'Lot Shape', 'Land Contour', 'Utilities', 'Lot Config', 'Land Slope', 'Neighborhood', 'Condition 1', 'Condition 2', 'Bldg Type', 'House Style', 'Roof Style', 'Roof Matl', 'Exterior 1st', 'Exterior 2nd', 'Mas Vnr Type', 'Exter Qual', 'Exter Cond', 'Foundation', 'Bsmt Qual', 'Bsmt Cond', 'Bsmt Exposure', 'BsmtFin Type 1', 'BsmtFin Type 2', 'Heating', 'Heating QC', 'Central Air', 'Electrical', 'Kitchen Qual', 'Functional', 'Fireplace Qu', 'Garage Type', 'Garage Finish', 'Garage Qual', 'Garage Cond', 'Paved Drive', 'Pool QC', 'Fence', 'Misc Feature', 'Sale Type', 'Sale Condition']

Numerical Columns:
['Order', 'PID', 'MS SubClass', 'Lot Frontage', 'Lot Area', 'Overall Qual', 'Overall Cond', 'Year Built', 'Year Remod/Add', 'Mas Vnr Area', 'BsmtFin SF 1', 'BsmtFin SF 2', 'Bsmt Unf SF', 'Total Bsmt SF', '1st Flr SF', '2nd Flr SF', 'Low Qual Fin SF', 'Gr Liv Area', 'Bsmt Full Bath', 'Bsmt Half Bath', 'Full Bath', 'Half Bath', 'Bedroom AbvGr

In [8]:
from sklearn.preprocessing import OneHotEncoder

# Instantiate OneHotEncoder
one_hot_encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)

# Fit and transform the categorical columns
X_categorical_encoded = one_hot_encoder.fit_transform(X[categorical_cols])

# Convert to DataFrame
X_categorical_encoded_df = pd.DataFrame(X_categorical_encoded, columns=one_hot_encoder.get_feature_names_out(categorical_cols))

display(X_categorical_encoded_df.head())

MS Zoning_A (agr)  MS Zoning_C (all)  MS Zoning_FV  MS Zoning_I (all)  \
0                0.0                0.0           0.0                0.0   
1                0.0                0.0           0.0                0.0   
2                0.0                0.0           0.0                0.0   
3                0.0                0.0           0.0                0.0   
4                0.0                0.0           0.0                0.0   

   MS Zoning_RH  MS Zoning_RL  MS Zoning_RM  Street_Grvl  Street_Pave  \
0           0.0           1.0           0.0          0.0          1.0   
1           1.0           0.0           0.0          0.0          1.0   
2           0.0           1.0           0.0          0.0          1.0   
3           0.0           1.0           0.0          0.0          1.0   
4           0.0           1.0           0.0          0.0          1.0   

   Alley_Grvl  ...  Sale Type_New  Sale Type_Oth  Sale Type_VWD  \
0         1.0  ...            0.0            0.0            0.0   
1         1.0  ...            0.0            0.0            0.0   
2         1.0  ...            0.0            0.0            0.0   
3         1.0  ...            0.0            0.0            0.0   
4         1.0  ...            0.0            0.0            0.0   

   Sale Type_WD   Sale Condition_Abnorml  Sale Condition_AdjLand  \
0            1.0                     0.0                     0.0   
1            1.0                     0.0                     0.0   
2            1.0                     0.0                     0.0   
3            1.0                     0.0                     0.0   
4            1.0                     0.0                     0.0   

   Sale Condition_Alloca  Sale Condition_Family  Sale Condition_Normal  \
0                    0.0                    0.0                    1.0   
1                    0.0                    0.0                    1.0   
2                    0.0                    0.0                    1.0   
3                    0.0                    0.0                    1.0   
4                    0.0                    0.0                    1.0   

   Sale Condition_Partial  
0                     0.0  
1                     0.0  
2                     0.0  
3                     0.0  
4                     0.0  

[5 rows x 267 columns]

In [9]:
from sklearn.preprocessing import StandardScaler

# Instantiate StandardScaler
scaler = StandardScaler()

# Fit and transform the numerical columns
X_numerical_scaled = scaler.fit_transform(X[numerical_cols])

# Convert to DataFrame
X_numerical_scaled_df = pd.DataFrame(X_numerical_scaled, columns=numerical_cols)

display(X_numerical_scaled_df.head())

Order       PID  MS SubClass  Lot Frontage  Lot Area  Overall Qual  \
0 -1.731460 -0.997164    -0.877005      3.366911  2.744381     -0.067254   
1 -1.730277 -0.996904    -0.877005      0.505463  0.187097     -0.776079   
2 -1.729095 -0.996899    -0.877005      0.552372  0.522814     -0.067254   
3 -1.727913 -0.996888    -0.877005      1.115280  0.128458      0.641571   
4 -1.726731 -0.992903     0.061285      0.224010  0.467348     -0.776079   

   Overall Cond  Year Built  Year Remod/Add  Mas Vnr Area  ...  Garage Area  \
0     -0.506718   -0.375537       -1.163488      0.056639  ...     0.256684   
1      0.393091   -0.342468       -1.115542     -0.571242  ...     1.196337   
2      0.393091   -0.441674       -1.259380      0.034215  ...    -0.748092   
3     -0.506718   -0.110988       -0.779919     -0.571242  ...     0.228774   
4     -0.506718    0.848000        0.658466     -0.571242  ...     0.042704   

   Wood Deck SF  Open Porch SF  Enclosed Porch  3Ssn Porch  Screen Porch  \
0      0.920121       0.214409       -0.358838   -0.103134     -0.285354   
1      0.366061      -0.704493       -0.358838   -0.103134      1.854530   
2      2.368594      -0.170937       -0.358838   -0.103134     -0.285354   
3     -0.742060      -0.704493       -0.358838   -0.103134     -0.285354   
4      0.935952      -0.200579       -0.358838   -0.103134     -0.285354   

   Pool Area   Misc Val   Mo Sold   Yr Sold  
0  -0.063031  -0.089422 -0.448057  1.678499  
1  -0.063031  -0.089422 -0.079602  1.678499  
2  -0.063031  21.985725 -0.079602  1.678499  
3  -0.063031  -0.089422 -0.816513  1.678499  
4  -0.063031  -0.089422 -1.184969  1.678499  

[5 rows x 38 columns]

In [10]:
X_processed = pd.concat([X_categorical_encoded_df, X_numerical_scaled_df], axis=1)
display(X_processed.head())

MS Zoning_A (agr)  MS Zoning_C (all)  MS Zoning_FV  MS Zoning_I (all)  \
0                0.0                0.0           0.0                0.0   
1                0.0                0.0           0.0                0.0   
2                0.0                0.0           0.0                0.0   
3                0.0                0.0           0.0                0.0   
4                0.0                0.0           0.0                0.0   

   MS Zoning_RH  MS Zoning_RL  MS Zoning_RM  Street_Grvl  Street_Pave  \
0           0.0           1.0           0.0          0.0          1.0   
1           1.0           0.0           0.0          0.0          1.0   
2           0.0           1.0           0.0          0.0          1.0   
3           0.0           1.0           0.0          0.0          1.0   
4           0.0           1.0           0.0          0.0          1.0   

   Alley_Grvl  ...  Garage Area  Wood Deck SF  Open Porch SF  Enclosed Porch  \
0         1.0  ...     0.256684      0.920121       0.214409       -0.358838   
1         1.0  ...     1.196337      0.366061      -0.704493       -0.358838   
2         1.0  ...    -0.748092      2.368594      -0.170937       -0.358838   
3         1.0  ...     0.228774     -0.742060      -0.704493       -0.358838   
4         1.0  ...     0.042704      0.935952      -0.200579       -0.358838   

   3Ssn Porch  Screen Porch  Pool Area   Misc Val   Mo Sold   Yr Sold  
0   -0.103134     -0.285354  -0.063031  -0.089422 -0.448057  1.678499  
1   -0.103134      1.854530  -0.063031  -0.089422 -0.079602  1.678499  
2   -0.103134     -0.285354  -0.063031  21.985725 -0.079602  1.678499  
3   -0.103134     -0.285354  -0.063031  -0.089422 -0.816513  1.678499  
4   -0.103134     -0.285354  -0.063031  -0.089422 -1.184969  1.678499  

[5 rows x 305 columns]

In [11]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_processed, y, test_size=0.2, random_state=42)

# Display the shapes of the resulting sets
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (2344, 305)
Shape of X_test: (586, 305)
Shape of y_train: (2344,)
Shape of y_test: (586,)


In [14]:
from sklearn.metrics import mean_squared_error, r2_score

# Evaluate the model using Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred)
print("Mean Squared Error (MSE):", mse)

# Evaluate the model using R-squared
r2 = r2_score(y_test, y_pred)
print("R-squared:", r2)

Mean Squared Error (MSE): 871728326.5315702
R-squared: 0.891272452120798


In [13]:
# Make predictions on the test data
y_pred = model.predict(X_test)

# Display the first few predictions
display(y_pred[:5])

array([155542.92420644, 110097.07540495, 200432.12271146, 129594.17439261,
       130835.79038672])

In [12]:
from sklearn.linear_model import LinearRegression

# Instantiate Linear Regression model
model = LinearRegression()

# Train the model
model.fit(X_train, y_train)

LinearRegression()

Here is a summary of the results:

Data Loading and Preprocessing: The Ames Housing dataset was loaded and missing values were handled by imputing categorical features with the mode and numerical features with the mean. Categorical features were then one-hot encoded, and numerical features were standardized. The data was split into training and testing sets.
Model Training: A Linear Regression model was trained on the preprocessed training data.
Prediction and Evaluation: Predictions were made on the test set, and the model's performance was evaluated using Mean Squared Error (MSE) and R-squared.
Mean Squared Error (MSE): 871,728,326.53
R-squared: 0.8913
The R-squared value of 0.8913 indicates that the model explains approximately 89.13% of the variance in house prices, suggesting a reasonably good fit to the data. The MSE provides a measure of the model's prediction error.

Would you like to explore any of the suggestions I mentioned earlier for improving the model?

# Task
Train and evaluate an XGBoost model and compare its performance with the Linear Regression model.

## Install xgboost

### Subtask:
Install the necessary library.


**Reasoning**:
The subtask is to install the xgboost library, which can be done using pip.



In [15]:
%pip install xgboost

## Train xgboost model

### Subtask:
Train an XGBoost model on the training data.


**Reasoning**:
Train an XGBoost model on the training data.



In [16]:
import xgboost as xgb

# Instantiate XGBoost Regressor model
xgb_model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)

# Train the model
xgb_model.fit(X_train, y_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             feature_weights=None, gamma=None, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=None, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=None,
             max_leaves=None, min_child_weight=None, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=None,
             n_jobs=None, num_parallel_tree=None, ...)

## Make predictions

### Subtask:
Make predictions on the test data using the trained XGBoost model.


**Reasoning**:
Use the trained XGBoost model to make predictions on the test data and display the first few predictions.



In [17]:
# Make predictions on the test data using the trained XGBoost model
y_pred_xgb = xgb_model.predict(X_test)

# Display the first few predictions
display(y_pred_xgb[:5])

array([176552.06,  97877.61, 198618.45, 124967.91, 114004.28],
      dtype=float32)

## Evaluate model

### Subtask:
Evaluate the performance of the XGBoost model using appropriate metrics (e.g., MSE, R-squared).


**Reasoning**:
Calculate and print the MSE and R-squared for the XGBoost model.



In [18]:
from sklearn.metrics import mean_squared_error, r2_score

# Calculate the Mean Squared Error (MSE) for the XGBoost model
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
print("Mean Squared Error (MSE) for XGBoost:", mse_xgb)

# Calculate the R-squared value for the XGBoost model
r2_xgb = r2_score(y_test, y_pred_xgb)
print("R-squared for XGBoost:", r2_xgb)

Mean Squared Error (MSE) for XGBoost: 609240320.0
R-squared for XGBoost: 0.9240116477012634


## Compare results

### Subtask:
Compare the performance of the Linear Regression model and the XGBoost model.


**Reasoning**:
Print the performance metrics for both models and then provide a brief comparison.



In [19]:
print("Linear Regression Model Performance:")
print("Mean Squared Error (MSE):", mse)
print("R-squared:", r2)
print("\nXGBoost Model Performance:")
print("Mean Squared Error (MSE):", mse_xgb)
print("R-squared:", r2_xgb)

print("\nModel Comparison:")
if r2_xgb > r2:
    print("The XGBoost model performed better than the Linear Regression model.")
    print(f"XGBoost has a higher R-squared ({r2_xgb:.4f} vs {r2:.4f}) and lower MSE ({mse_xgb:.2f} vs {mse:.2f}).")
elif r2_xgb < r2:
    print("The Linear Regression model performed better than the XGBoost model.")
    print(f"Linear Regression has a higher R-squared ({r2:.4f} vs {r2_xgb:.4f}) and lower MSE ({mse:.2f} vs {mse_xgb:.2f}).")
else:
    print("Both models performed similarly.")


Linear Regression Model Performance:
Mean Squared Error (MSE): 871728326.5315702
R-squared: 0.891272452120798

XGBoost Model Performance:
Mean Squared Error (MSE): 609240320.0
R-squared: 0.9240116477012634

Model Comparison:
The XGBoost model performed better than the Linear Regression model.
XGBoost has a higher R-squared (0.9240 vs 0.8913) and lower MSE (609240320.00 vs 871728326.53).


## Summary:

### Data Analysis Key Findings

*   The XGBoost model achieved a Mean Squared Error (MSE) of approximately 609,240,320.0 on the test data.
*   The R-squared value for the XGBoost model on the test data is approximately 0.924.
*   Compared to the Linear Regression model (MSE of approximately 871,728,326.53 and R-squared of approximately 0.8913), the XGBoost model shows better performance with a lower MSE and a higher R-squared.

### Insights or Next Steps

*   The XGBoost model is a promising candidate for this regression task, outperforming the baseline Linear Regression model.
*   Further hyperparameter tuning of the XGBoost model could potentially improve its performance even further.
